pootle/apps/virtualfolder/receivers.py

```diff
@@ -0,0 +1,133 @@
+# AUTHORS file for copyright and authorship information.
+
+from django.db.models.signals import post_save, pre_save
+from django.dispatch import receiver
+
+from pootle_store.models import Unit
+from pootle_store.util import OBSOLETE
+
+from .models import VirtualFolder, VirtualFolderTreeItem
+from .signals import vfolder_post_save
+
+
+def update_vfolder_tree(vf, store):
+    # Create missing VirtualFolderTreeItem tree structure after
+    # adding this new unit.
+    vfolder_treeitem, created = (
+        VirtualFolderTreeItem.objects.get_or_create(
+            directory=store.parent, vfolder=vf))
+
+    if not created:
+        # The VirtualFolderTreeItem already existed, so
+        # calculate again the stats up to the root.
+        vfolder_treeitem.update_all_cache()
+
+
+def add_unit_to_vfolders(unit):
+    pootle_path = unit.store.pootle_path
+
+    for vf in VirtualFolder.objects.iterator():
+        unit_added = False
+        for location in vf.all_locations:
+            if not pootle_path.startswith(location):
+                continue
+
+            for filename in vf.filter_rules.split(","):
+                if pootle_path == "".join([location, filename]):
+                    vf.units.add(unit)
+                    unit_added = True
+                    break
+
+            if unit_added:
+                break
+
+        if unit_added:
+            update_vfolder_tree(vf, unit.store)
+
+
+@receiver(pre_save, sender=VirtualFolder)
+def vfolder_unit_priority_presave_handler(sender, instance, **kwargs):
+
+    if instance.id is None:
+        return
+
+    original = VirtualFolder.objects.get(pk=instance.pk)
+
+    original_locations = set()
+    new_locations = set()
+
+    for location in original.all_locations:
+        for filename in original.filter_rules.split(","):
+            original_locations.add("".join([location, filename]))
+
+    for location in instance.all_locations:
+        for filename in instance.filter_rules.split(","):
+            new_locations.add("".join([location, filename]))
+
+    removed_locations = original_locations - new_locations
+
+    stores_affected = set()
+    for location in removed_locations:
+        # reindex these units without this vfolder
+        removed_units = (
+            Unit.objects.filter(store__pootle_path__startswith=location,
+                                vfolders=original))
+        for unit in removed_units.iterator():
+            unit.vfolders.remove(original)
+            unit.set_priority()
+            stores_affected.add(unit.store)
+
+    for store in stores_affected:
+        update_vfolder_tree(original, store)
+
+
+@receiver(vfolder_post_save, sender=VirtualFolder)
+def vfolder_unit_priority_handler(sender, instance, **kwargs):
+
+    for unit in instance.units.all():
+        unit.set_priority()
+
+
+@receiver(pre_save, sender=Unit)
+def vfolder_unit_unobsoleted(sender, instance, created=False, **kwargs):
+    if instance.state == OBSOLETE:
```

Comment:
This is pre_save for unobsolete. So why don't we do anything if unit is obsolete?